### PR TITLE
add solution

### PR DESCRIPTION
**Pull request by Victoria-Moiseienko**

### CODE REVIEW

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+            PreparedStatement preparedStatement =
+                    connection.prepareStatement(query, Statement.RETURN_GENERATED_KEYS);
+            preparedStatement.setString(1, item.getName());
+            preparedStatement.setDouble(2, item.getPrice());
+            preparedStatement.executeUpdate();
+            ResultSet resultKey = preparedStatement.getGeneratedKeys();
+            if (resultKey.next()) {
+                item.setId(resultKey.getLong(1));
+            }
+
```
Comment (TimoxaV) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: empty line

**Review by TimoxaV — APPROVED** [review]

File: src/main/java/com/internet/shop/TestDataBaseWork.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+public class TestDataBaseWork {
+    public static void main(String[] args) {
+        ProductDao dao = new ProductDaoJdbcImpl();
+        //insert
+        Product testItem = dao.create(new Product("mango", 99));
+        //select
+        dao.get(testItem.getId()).ifPresent(product -> System.out.println(product.toString()));
+        //update
+        testItem.setPrice(18);
+        dao.update(testItem);
```
Comment (boroda4436) [inline src/main/java/com/internet/shop/TestDataBaseWork.java:17]: would be nice to print data after the update
```suggestion
        dao.update(testItem);
        dao.get(testItem.getId()).ifPresent(product -> System.out.println(product.toString()));
```

File: src/main/java/com/internet/shop/TestDataBaseWork.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+        //delete
+        dao.delete(testItem.getId());
+        //select all
+        List<Product> all = dao.getAll();
```
Comment (boroda4436) [inline src/main/java/com/internet/shop/TestDataBaseWork.java:22]: I guess there will be an empty list.

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java (hunk 1 above)
Comment (boroda4436) [reply src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: @Victoria-Moiseienko WDYT?

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+public class ProductDaoJdbcImpl implements ProductDao {
+    @Override
+    public Product create(Product item) {
```
Comment (boroda4436) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]:
```suggestion
    public Product create(Product product) {
```

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+        } catch (SQLException e) {
+            throw new DataProcessingException("Item has not been added/n" + e);
+        }
+        return item;
+    }
+
+    @Override
+    public Optional<Product> get(Long id) {
+        String query = "SELECT product_id, product_name, product_price"
+                + " FROM products WHERE product_id = ?";
```
Comment (boroda4436) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]:
```suggestion
                + " FROM products WHERE product_id = ? and deleted = false";
```

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+        try (Connection connection = DbConnectionUtil.getConnection()) {
+            PreparedStatement preparedStatement = connection.prepareStatement(query);
+            preparedStatement.setLong(1, id);
+            ResultSet sqlResult = preparedStatement.executeQuery();
+            if (sqlResult.next()) {
+                Product product = parseRow(sqlResult);
+                return Optional.of(product);
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Item has not been selected/n" + e);
```
Comment (boroda4436) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]:
```suggestion
            throw new DataProcessingException("Can't get product by id = " + id, e);
```

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+            preparedStatement.setString(1, item.getName());
+            preparedStatement.setDouble(2, item.getPrice());
+            preparedStatement.executeUpdate();
+            ResultSet resultKey = preparedStatement.getGeneratedKeys();
+            if (resultKey.next()) {
+                item.setId(resultKey.getLong(1));
+            }
+
+        } catch (SQLException e) {
+            throw new DataProcessingException("Item has not been added/n" + e);
```
Comment (boroda4436) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]:
```suggestion
            throw new DataProcessingException("Item has not been added", e);
```
do not concatenate message and exception object

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+            throw new DataProcessingException("Item has not been selected/n" + e);
+        }
+        return Optional.empty();
+    }
+
+    @Override
+    public Product update(Product item) {
+        String query = "UPDATE products"
+                + " SET product_name = ?, product_price = ? "
+                + "WHERE product_id = ?";
```
Comment (boroda4436) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java:62]: + deleted

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+
+        try (Connection connection = DbConnectionUtil.getConnection()) {
+            PreparedStatement prepareStatement = connection.prepareStatement(query);
+            prepareStatement.setString(1, item.getName());
+            prepareStatement.setDouble(2, item.getPrice());
+            prepareStatement.setLong(3, item.getId());
+            prepareStatement.executeUpdate();
+        } catch (SQLException e) {
+            throw new DataProcessingException("Item has not been updated/n" + e);
```
Comment (boroda4436) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: same

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+        String query = "UPDATE products"
+                + " SET deleted = 1 "
+                + "WHERE product_id = ?";
+
+        try (Connection connection = DbConnectionUtil.getConnection()) {
+            PreparedStatement prepareStatement = connection.prepareStatement(query);
+            prepareStatement.setLong(1, id);
+            prepareStatement.executeUpdate();
+        } catch (SQLException e) {
+            throw new DataProcessingException("Item has not been deleted/n" + e);
```
Comment (boroda4436) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: .

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+        } catch (SQLException e) {
+            throw new DataProcessingException("Item has not been updated/n" + e);
+        }
+        return item;
+    }
+
+    @Override
+    public boolean delete(Long id) {
+        String query = "UPDATE products"
+                + " SET deleted = 1 "
```
Comment (boroda4436) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]:
```suggestion
                + " SET deleted = true "
```

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+                + "WHERE product_id = ?";
```
Comment (boroda4436) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]:
```suggestion
                + "WHERE product_id = ? and deleted = false";
```

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+        } catch (SQLException e) {
+            throw new DataProcessingException("Item has not been deleted/n" + e);
+        }
+        return true;
+    }
+
+    @Override
+    public List<Product> getAll() {
+        String query = "SELECT product_id, product_name, product_price"
+                + " FROM products WHERE deleted = 0";
```
Comment (boroda4436) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]:
```suggestion
                + " FROM products WHERE deleted = false";
```

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+            prepareStatement.executeUpdate();
+        } catch (SQLException e) {
+            throw new DataProcessingException("Item has not been deleted/n" + e);
+        }
+        return true;
+    }
+
+    @Override
+    public List<Product> getAll() {
+        String query = "SELECT product_id, product_name, product_price"
```
Comment (boroda4436) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]:
```suggestion
        String query = "SELECT product_id, name, price"
```

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+                + " FROM products WHERE deleted = 0";
+        List<Product> productList = new ArrayList<>();
+
+        try (Connection connection = DbConnectionUtil.getConnection()) {
+            PreparedStatement preparedStatement = connection.prepareStatement(query);
+            ResultSet sqlResult = preparedStatement.executeQuery();
+
+            while (sqlResult.next()) {
+                Product product = parseRow(sqlResult);
+                productList.add(product);
```
Comment (boroda4436) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]:
```suggestion
                productList.add(parseRow(sqlResult));
```

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Items has not been selected/n" + e);
```
Comment (boroda4436) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: .

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+        }
+        return productList;
+    }
+
+    private Product parseRow(ResultSet sqlResult) throws SQLException {
+        Long id = sqlResult.getLong("product_id");
+        String name = sqlResult.getString("product_name");
```
Comment (boroda4436) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]:
```suggestion
        String name = sqlResult.getString("name");
```

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 18]
```diff
@@ -0,0 +1,118 @@
...
+        Double price = sqlResult.getDouble("product_price");
```
Comment (boroda4436) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]:
```suggestion
        Double price = sqlResult.getDouble("price");
```

File: src/main/java/com/internet/shop/util/DbConnectionUtil.java [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+            Class.forName("com.mysql.jdbc.Driver");
+        } catch (ClassNotFoundException e) {
+            throw new RuntimeException("Can't find MySQL Driver", e);
+        }
+    }
+
+    public static Connection getConnection() throws SQLException {
+        Properties connectionProps = new Properties();
+        connectionProps.put("user", "root");
+        connectionProps.put("password", "Vi4ft5q7Ktrr9T6toPcN");
```
Comment (boroda4436) [inline src/main/java/com/internet/shop/util/DbConnectionUtil.java]: do not push your real credentials to GutHub. Never

File: src/main/java/com/internet/shop/util/DbConnectionUtil.java [hunk 20]
```diff
@@ -0,0 +1,29 @@
...
+        String url = "jdbc:mysql://localhost:3306/internet_shop?serverTimezone=UTC";
+
+        try {
+            return DriverManager.getConnection(url, connectionProps);
+        } catch (SQLException e) {
+            throw new RuntimeException("Can't establish the connection to DB:/n" + e);
```
Comment (boroda4436) [inline src/main/java/com/internet/shop/util/DbConnectionUtil.java]: .

File: src/main/resources/init_db.sql [hunk 21]
```diff
@@ -0,0 +1,16 @@
+CREATE SCHEMA `internet_shop` DEFAULT CHARACTER SET utf8 ;
+
+CREATE TABLE `internet_shop`.`products` (
+  `product_id` BIGINT(11) NOT NULL,
+  `product_name` VARCHAR(225) NOT NULL,
+  `product_price` DOUBLE NULL,
+  PRIMARY KEY (`product_id`),
+  UNIQUE INDEX `products_id_UNIQUE` (`product_id` ASC) VISIBLE,
+  UNIQUE INDEX `product_name_UNIQUE` (`product_name` ASC) VISIBLE)
```
Comment (boroda4436) [inline src/main/resources/init_db.sql:9]: Why product name is unique?

**Review by boroda4436 — CHANGES_REQUESTED** [review]

File: src/main/resources/init_db.sql (hunk 21 above)
Comment (Victoria-Moiseienko) [reply src/main/resources/init_db.sql:9]: It's error - I know it after morning lection

File: src/main/java/com/internet/shop/util/DbConnectionUtil.java (hunk 20 above)
Comment (Victoria-Moiseienko) [reply src/main/java/com/internet/shop/util/DbConnectionUtil.java]: Done

File: src/main/java/com/internet/shop/util/DbConnectionUtil.java (hunk 19 above)
Comment (Victoria-Moiseienko) [reply src/main/java/com/internet/shop/util/DbConnectionUtil.java]: Ok. Done

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java (hunk 18 above)
Comment (Victoria-Moiseienko) [reply src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: Done

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java (hunk 17 above)
Comment (Victoria-Moiseienko) [reply src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: Done

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java (hunk 1 above)
Comment (Victoria-Moiseienko) [reply src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: Done

File: src/main/java/com/internet/shop/TestDataBaseWork.java (hunk 2 above)
Comment (Victoria-Moiseienko) [reply src/main/java/com/internet/shop/TestDataBaseWork.java:17]: Done

File: src/main/java/com/internet/shop/TestDataBaseWork.java (hunk 3 above)
Comment (Victoria-Moiseienko) [reply src/main/java/com/internet/shop/TestDataBaseWork.java:22]: Products in DB contains some rows before start of testing :)

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java (hunk 4 above)
Comment (Victoria-Moiseienko) [reply src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: Done

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java (hunk 5 above)
Comment (Victoria-Moiseienko) [reply src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: Done

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java (hunk 11 above)
Comment (Victoria-Moiseienko) [reply src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: Done

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java (hunk 15 above)
Comment (Victoria-Moiseienko) [reply src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: Done

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java (hunk 14 above)
Comment (Victoria-Moiseienko) [reply src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: Done

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java (hunk 13 above)
Comment (Victoria-Moiseienko) [reply src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: Done

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java (hunk 12 above)
Comment (Victoria-Moiseienko) [reply src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: Done

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java (hunk 10 above)
Comment (Victoria-Moiseienko) [reply src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: Done

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java (hunk 9 above)
Comment (Victoria-Moiseienko) [reply src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: Done

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java (hunk 8 above)
Comment (Victoria-Moiseienko) [reply src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java:62]: Done

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java (hunk 7 above)
Comment (Victoria-Moiseienko) [reply src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: Done

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java (hunk 6 above)
Comment (Victoria-Moiseienko) [reply src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: Done

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 22]
```diff
@@ -0,0 +1,117 @@
...
+                    connection.prepareStatement(query, Statement.RETURN_GENERATED_KEYS);
+            preparedStatement.setString(1, product.getName());
+            preparedStatement.setDouble(2, product.getPrice());
+            preparedStatement.executeUpdate();
+            ResultSet resultKey = preparedStatement.getGeneratedKeys();
+            if (resultKey.next()) {
+                product.setId(resultKey.getLong(1));
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Item has not been added", e);
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]:
```suggestion
            throw new DataProcessingException("Product with name " + name + " has not been added", e);
```

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 23]
```diff
@@ -0,0 +1,117 @@
...
+
+        try (Connection connection = DbConnectionUtil.getConnection()) {
+            PreparedStatement prepareStatement = connection.prepareStatement(query);
+            prepareStatement.setString(1, product.getName());
+            prepareStatement.setDouble(2, product.getPrice());
+            prepareStatement.setBoolean(3, true);
+            prepareStatement.setLong(4, product.getId());
+            prepareStatement.executeUpdate();
+        } catch (SQLException e) {
+            throw new DataProcessingException("Item has not been updated", e);
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]:
```suggestion
            throw new DataProcessingException("Product with id " + product.getId() + " has not been updated", e);
```

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 24]
```diff
@@ -0,0 +1,117 @@
...
+        String query = "UPDATE products"
+                + " SET deleted = true "
+                + "WHERE product_id = ? and deleted = false";
+
+        try (Connection connection = DbConnectionUtil.getConnection()) {
+            PreparedStatement prepareStatement = connection.prepareStatement(query);
+            prepareStatement.setLong(1, id);
+            prepareStatement.executeUpdate();
+        } catch (SQLException e) {
+            throw new DataProcessingException("Item has not been deleted", e);
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: same here

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 25]
```diff
@@ -0,0 +1,117 @@
...
+
+        try (Connection connection = DbConnectionUtil.getConnection()) {
+            PreparedStatement preparedStatement = connection.prepareStatement(query);
+            ResultSet sqlResult = preparedStatement.executeQuery();
+
+            while (sqlResult.next()) {
+                productList.add(parseRow(sqlResult));
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Items has not been selected", e);
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: fix this message

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java (hunk 22 above)
Comment (Victoria-Moiseienko) [reply src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: Done

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java (hunk 23 above)
Comment (Victoria-Moiseienko) [reply src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: Done

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java (hunk 24 above)
Comment (Victoria-Moiseienko) [reply src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: Done

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java (hunk 25 above)
Comment (Victoria-Moiseienko) [reply src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: Done

**Review by RomanDubovskyi — APPROVED** [review]

**Pull request merged**